### PR TITLE
[GPII-3868]: Gracefully rotate deployment credentials

### PR DESCRIPTION
**Pull request by natarajaya**
This is a proof-of-concept solution that gracefully rotates SA credentials managed in TF (https://github.com/gpii-ops/gpii-infra/pull/429).
It needs new `kubectl` that is capable of doing rolling restarts (https://github.com/gpii-ops/exekube/pull/59).

### CODE REVIEW

File: shared/rakefiles/xk_util.rake [hunk 1]
```diff
@@ -55,6 +55,34 @@ task :rotate_secrets_key, [:encryption_key] => [:configure, :configure_secrets]
...
+  sh "kubectl -n #{namespace} rollout restart deployment #{args[:deployment]}"
+
+  # Rollout status ensures that all pods updated (waiting)
+  sh "kubectl -n #{namespace} rollout status deployment #{args[:deployment]}"
+
+  # Destroying existing credentials
+  sh "rake destroy_sa_keys['gke-cluster-pod-#{args[:deployment]}']"
+
+  # Applying deployment module to force TF to generate new credentials and re-create K8s secret
+  sh "rake xk['apply live/#{@env}/k8s/#{namespace}/#{args[:deployment]}',true,false,true]"
```
Comment (natarajaya) [inline shared/rakefiles/xk_util.rake]: This is just simplest tool for the purpose I can come up with. We may later replace this with `gcloud` command that generates the new keys and then imports updated resource into TF state, to not "modify the state of the world".

**Comment by amatas** [comment]
Is it possible to destroy the old keys only when the pods are using the new ones?, my concern here is that would be some kind of disruption between the keys are destroyed and when the pods are restarted to get the new credentials.

**Comment by natarajaya** [comment]
> Is it possible to destroy the old keys only when the pods are using the new ones?

@amatas Keys are not being used all the time. On startup, `trace-agent` would use the keys to generate token, which is [valid for 1 hour](https://cloud.google.com/iam/docs/creating-short-lived-service-account-credentials#sa-credentials-oauth). Token remains valid even if the key is destroyed, so we can safely re-generate new credentials and rolling restart deployments.

**Comment by stepanstipl** [comment]
So this is something I wanted to avoid - investing time in implementing high-level logic to fix some low-level issue, non-portable and tightly coupled with rake code and our deployment process.

Seems the predicate our previous conversation was based on, that the app will notice the change of credentials without restart, is not true. But this simply means that should we want to rotate the credentials, we have to restart the pods - yet this can still be done in or outside of the cluster.

Some of the issues:
- Tightly coupled with our deploy process - if we don't run deployments (let's say it Christmas break, or simply in future the project is stable that it doesn't need changes) the credentials won't be rotated
- Static credentials are pulled to the CI and stored in Terraform state, as opposed to never leaving cluster/GCP (or not having static credentials at all)
- Not portable and conceptually in the wrong place - this does not belong to the rake code - if we use Helm as a packaging format for our applications, this should be handled there
- Obviously having to restart the application because of rotating tracing credentials does not seem to be very nice
- An issue with dependency on the unsupported version of `kubectl` - this can be worked-around as mentioned in gpii-ops/exekube#59
- Complexity - instead of fixing the underlying issue, we're introducing more code (== read more opportunities for failure)

> Keys are not being used all the time. On startup, trace-agent would use the keys to generate token, which is valid for 1 hour. Token remains valid even if the key is destroyed, so we can safely re-generate new credentials and rolling restart deployments.

Also there will be a short window where the existing key is disabled, but the pods have not been updated and restarted. If the app would try to get a new token in that time, I assume it would fail.


For all of the above, my preferences for any other than short temporary solution would be:
1. Fix the underlying issue (btw. I would like to know how the tracing behaves with the new GKE Workload Identity)

2. If not, then Implement this in-cluster, my idea is a Cron job that
  - Uses `gcloud` via metadata credentials to generate the new svc account key that the app is able to consume
  - Uploads this key to kubernetes secret
  - Uploads some annotation on the deployment to trigger a restart of all the apps (I really dislike the restart part)
  - Revokes the old key

  While this wouldn't avoid the restart, the solution would be self-contained in the Helm Chart, the credentials would never leave cluster, we wouldn't need to modify rake code and the key rotation would reliably happen at predictable interval. Downside is we would need a docker image containing both `gcloud` and `kubectl` (I don't think we have one atm.).

**Comment by natarajaya** [comment]
> So this is something I wanted to avoid - investing time in implementing high-level logic to fix some low-level issue

@stepanstipl Well, as you know, low-level issue is not something that we can fix ourselves. So far I've invested 1 hour of my time into this proof-of-concept and already have working solution that does the job (I, honestly, spent more time in this argument than working on this implementation).

> Tightly coupled with our deploy process - if we don't run deployments (let's say it Christmas break, or simply in future the project is stable that it doesn't need changes) the credentials won't be rotated

Pipeline runs nightly and goes all the way to prod regardless of the fact that there may be no new changes to deploy.

> Static credentials are pulled to the CI and stored in Terraform state, as opposed to never leaving cluster/GCP (or not having static credentials at all)

They are handled entirely by TF, never written to disk, TF state is encrypted.

> Not portable and conceptually in the wrong place - this does not belong to the rake code - if we use Helm as a packaging format for our applications, this should be handled there

I am not sure why it has anything to do with rake or Helm. 

> Obviously having to restart the application because of rotating tracing credentials does not seem to be very nice

Do you have an alternative that does not require pod restarts?

> An issue with dependency on the unsupported version of kubectl

Just a simplest tool for the purpose that I used to provide this proof-of-concept. We can keep `kubectl1.15` only to do rollout restarts or switch to another less convenient solution.

> Complexity 

But alternative solution that you are proposing look even more complex: in addition to the same credentials rotation logic that just moves to different level, we'll need to build and maintain separate image for sidecar (or cron, whatever it is), need to monitor that it is actually running and credentials are being rotated.

> instead of fixing the underlying issue

Do you have a proposition on how to fix the underlying issue? If not - then what is the point of this argument?

> Also there will be a short window where the existing key is disabled, but the pods have not been updated and restarted. If the app would try to get a new token in that time, I assume it would fail.

Thats why this proof-of-concept does rollout restart twice: first – to make sure that all pods generated new tokens, second – so new pods can use updated credentials.

> Fix the underlying issue (btw. I would like to know how the tracing behaves with the new GKE Workload Identity)

This is actually a good idea. Let me try this – looks like it can be enabled on current GKE version, so there is no need to upgrade to non-default 1.13.

> If not, then Implement this in-cluster, my idea is a Cron job that

Sorry, but having a cron that constantly restarts pods in the cluster sounds like a pretty bad idea to me.

**Comment by stepanstipl** [comment]
Thanks @natarajaya for replies.

> @stepanstipl Well, as you know, low-level issue is not something that we can fix ourselves. So far I've invested 1 hour of my time into this proof-of-concept and already have working solution that does the job (I, honestly, spent more time in our argument than working on this implementation).

I get that, however, time spent on a solution does not proportionally correlate with its quality.

This PoC would potentially restart both preferences up to 3 times in one deployment:
- Once when the image is actually updated as part of `rake` run
- 2nd & 3rd time when the `rake rotate_deployment_credentials` runs

I did a quick test in my dev cluster and one restart of preferences took 3m:
```
Fri  5 Jul 2019 12:02:20 BST
deployment.extensions/preferences restarted
Waiting for deployment "preferences" rollout to finish: 1 out of 2 new replicas have been updated...
Waiting for deployment "preferences" rollout to finish: 1 out of 2 new replicas have been updated...
Waiting for deployment "preferences" rollout to finish: 1 out of 2 new replicas have been updated...
Waiting for deployment "preferences" rollout to finish: 1 old replicas are pending termination...
Waiting for deployment "preferences" rollout to finish: 1 old replicas are pending termination...
deployment "preferences" successfully rolled out
Fri  5 Jul 2019 12:05:04 BST
```

Now we have to do this 2x, for both preferences & flowmanager, so **this solution adds 12m** to every pipeline run, probably more (I'm not counting the other steps and deployments in stg/prd have more pods.. btw how does this approach scale?).

> Pipeline runs nightly and goes all the way to prod regardless of the fact that there may be no new changes to deploy.

The pipeline needs 2 manual promotions to run in prod. To illustrate this on last week - pipeline in prd ran on Jul 28th, 9:34pm and then next time on Jul 2nd, 2L34pm - 4 days gap.  So if there are no manual promotions -> no runs in prod -> no rotation.

> I am not sure why it has anything to do with rake or Helm.

Everything the app needs to operate should be in the app package if possible, in our case Helm -> not on other layers of our stack.

> Sorry, but having a cron that constantly restarts pods in the cluster sounds like a pretty bad idea to me.

Can you elaborate on how is this worse than restarting all pods in workload up to 3 times on deployment, with unpredictable intervals between credentials rotation and adding a substantial amount of time to our pipeline?

**Comment by stepanstipl** [comment]
Anyway, I think this whole discussion is now not really relevant anymore:

I did try to enable [GKE Workload Identity](https://cloud.google.com/kubernetes-engine/docs/how-to/workload-identity) on 1.12 cluster and tracing seems to work out of the box.

```
$ gcloud auth list
Credentialed Accounts
ACTIVE  ACCOUNT
*       gke-cluster-pod-default@gpii-gcp-dev-stepan.iam.gserviceaccount.com
```

Screenshot 2019-07-05 at 14 00 26

So my proposal here is to:
- Drop workarounds with static credentials injection to pods
- Get rid of service-account-assigner
- Enable and migrate to GKE Workload identity

This will give us a Google-managed & supported solution that works without any hacks out-of-the-box with `google-cloud/trace-agent` and our apps. The only downside I see is that it's in beta, but I would go for it.

**Comment by natarajaya** [comment]
> I get that, however, time spent on a solution does not proportionally correlate with its quality.

@stepanstipl This is a proof-of-concept that with minimum effort (and minimal downside) improves security of an ugly temporary workaround for metadata issues. Why exactly are we talking about quality here?

> so this solution adds 12m to every pipeline run, probably more (I'm not counting the other steps and deployments in stg/prd have more pods.. btw how does this approach scale?)

From my observations – it takes from 30 seconds to 3 minutes to complete rollout restart in prod-like environment. This adds extra 5-8% to the total pipeline duration. Regarding scalability: rollout duration does not necessarily increase with the number of pods, it mostly depends on rolling update strategy and amount of available resources in the cluster. Also, any solution that requires pod restarts will have to face similar issues.

> To illustrate this on last week - pipeline in prd ran on Jul 28th, 9:34pm and then next time on Jul 2nd, 2L34pm - 4 days gap

I think this is pretty much worst case scenario – 4 days without rotation. How often we rotate CI keys atm?

> Everything the app needs to operate should be in the app package if possible, in our case Helm -> not on other layers of our stack.

Yes, and normally our apps should be able to retrieve access tokens using metadata API and does not require static credentials. We are dealing with workaround here – I don't understand why higher levels of the stack (that are already involved in secret management) can't be involved here.

> Can you elaborate on how is this worse

From the top of my head:
* Credential rotation and pod restarts happen when we expect them (on pipeline run) and actively monitor systems VS everything happens in background periodically.
* In case of any failure, we get notified immediately (failed pipeline) VS something fails quietly in background and requires additional monitoring.
* Smoke tests are being executed after rotation and restarts to make sure that all systems are functional VS we rely on K8s entirely and hope that nothing goes wrong during background rollouts, opening new potential ways for a failure.

**Comment by natarajaya** [comment]
> I did try to enable GKE Workload Identity on 1.12 cluster and tracing seems to work out of the box.

@stepanstipl Thanks for giving this a try! I got same result from my experiments, however, it turned out that it has nothing to do with Workload Identity, and the real culprit here is `service-account-assigner`! :(

With Workload Identity disabled, I was able to get proper SA token responses from metadata API after changing `disable-legacy-endpoints` to `false` on node pool and destroying `service-account-assigner` proxy.

> Drop workarounds with static credentials injection to pods Get rid of service-account-assigner Enable and migrate to GKE Workload identity

This plan looks reasonable, though I am a little worried that migration to Beta Workload Identity may bring another complications. As alternative that does not require migration from existing SA workflow, we can try to fix `service-account-assigner` [server](https://github.com/imduffy15/k8s-gke-service-account-assigner/blob/master/server/server.go#L248) to respond with proper `Metadata-Flavor` header on token requests.

**Comment by natarajaya** [comment]
Closing this, since there is now a fix for `service-account-assigner` that solves the metadata problem.

**Comment by stepanstipl** [comment]
thanks @natarajaya.

> since there is now a fix for service-account-assigner

Can you please link to the fix?

**Comment by natarajaya** [comment]
@stepanstipl Sure: https://github.com/gpii-ops/k8s-gke-service-account-assigner/pull/2

**Pull request closed without merging**